service/history/ndc/workflow_state_replicator.go

```diff
@@ -187,7 +198,551 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
+	ctx context.Context,
+	namespaceId namespace.ID,
+	workflowId string,
+	runId string,
+	sourceClusterName string,
+	wfCtx workflow.Context,
+	localMutableState workflow.MutableState,
+	sourceVersionHistories *history.VersionHistories,
+	eventBlobs []*commonpb.DataBlob,
+) error {
+	sourceVersionHistory, err := versionhistory.GetCurrentVersionHistory(sourceVersionHistories)
+	if err != nil {
+		return err
+	}
+	localVersionHistories := localMutableState.GetExecutionInfo().GetVersionHistories()
+	index, isNewBranch, err := func() (int32, bool, error) {
+		lcaItem, index, err := versionhistory.FindLCAVersionHistoryItemAndIndex(localMutableState.GetExecutionInfo().VersionHistories, sourceVersionHistory)
+		if err != nil {
+			return 0, false, err
+		}
+
+		localHistory, err := versionhistory.GetVersionHistory(localVersionHistories, index)
+		if err != nil {
+			return 0, false, err
+		}
+		if versionhistory.IsLCAVersionHistoryItemAppendable(localHistory, lcaItem) {
+			return index, false, nil
+		}
+		newVersionHistory, err := versionhistory.CopyVersionHistoryUntilLCAVersionHistoryItem(localHistory, lcaItem)
+		if err != nil {
+			return 0, false, err
+		}
+		branchManager := NewBranchMgr(r.shardContext, wfCtx, localMutableState, r.logger)
+		newVersionHistoryIndex, err := branchManager.createNewBranch(
+			ctx,
+			localHistory.GetBranchToken(),
+			lcaItem.GetEventId(),
+			newVersionHistory,
+		)
+		if err != nil {
+			return 0, false, err
+		}
 
+		return newVersionHistoryIndex, true, nil
+	}()
+
+	if err != nil {
+		return err
+	}
+	versionHistoryToAppend, err := versionhistory.GetVersionHistory(localVersionHistories, index)
+	if err != nil {
+		return err
+	}
+	localLastItem, err := versionhistory.GetLastVersionHistoryItem(versionHistoryToAppend)
+	if err != nil {
+		return err
+	}
+	sourceLastItem, err := versionhistory.GetLastVersionHistoryItem(sourceVersionHistory)
+	if err != nil {
+		return err
+	}
+	if versionhistory.IsEqualVersionHistoryItem(localLastItem, sourceLastItem) {
+		return nil
+	}
+
+	startEventId := localLastItem.GetEventId()
+	startEventVersion := localLastItem.GetVersion()
+	endEventId := sourceLastItem.GetEventId()
+	endEventVersion := sourceLastItem.GetVersion()
+	var historyEvents [][]*historypb.HistoryEvent
+	for _, blob := range eventBlobs {
+		events, err := r.historySerializer.DeserializeEvents(blob)
+		if err != nil {
+			return err
+		}
+		historyEvents = append(historyEvents, events)
+	}
+
+	prevTxnID := common.EmptyEventTaskID
+	fetchFromRemoteAndAppend := func(
+		startId, // exclusive
+		startVersion,
+		endId, // inclusive
+		endVersion int64) error {
+		remoteHistoryIterator := collection.NewPagingIterator(r.getHistoryFromRemotePaginationFn(
+			ctx,
+			sourceClusterName,
+			namespaceId,
+			workflowId,
+			runId,
+			startId,
+			startVersion,
+			endId,
+			endVersion),
+		)
+		for remoteHistoryIterator.HasNext() {
+			historyBlob, err := remoteHistoryIterator.Next()
```

Comment:
Is one blob one history batch? How many persistence requests will be made in existing resender impl v.s this impl.

Reply:
Yeah, seems this is also used by current resend. The current closed workflow replication back fill logic is also using the same logic.

 I did not find an API that writes multiple batches of history events.